### PR TITLE
filestate/upgrade: Add hook to fill missing project names

### PR DESCRIPTION
**Pull request by abhinav**
Commits are individually reviewable.

---

The filestate upgrade operation is limited: it can only upgrade stacks
for which it's able to guess a project name by inspecting resources.
It cannot upgrade empty stacks, or stacks where someone recently called
`pulumi destroy` because they don't have any resources for it to inspect.

To resolve this, we need to support an external hook:
some means for the caller to fill in project names for stacks
where the names could not be guessed.
We'll use this hook in the CLI to prompt the user.

This change adds an UpgradeOptions argument to the Upgrade operation.
UpgradeOptions includes an option ProjectsForDetachedStacks.
This is a callback in the form:

    func(stacks []Name) (projects []Name, err error)

It takes a list of stack names and returns a list of project names for them
in the same order.

This also has the effect of splitting the upgrade operation.
Previously, it was roughly:

    for _, s := range stacks {
      go func(s) {
        proj := guessProject()
        if proj == "" { skip }
        upgrade(s, proj)
      }(s)
    }

A single parallelizable unit of work.

With this change, we have three separate steps. Roughly:

    for _, s := range stacks {
      go func(s) {
        guessProject(s)
        // ...
      }(s)
    }

    fillMissingProjects(...)

    for _, s := range stacks {
      go upgrade(s, projects[s])
    }

Only the first and third step are run concurrently.
The fillMissingProjects is blocking and runs after the first phase
has finished completely.

This change leverages the abstraction introduced in #13076
to make the above more manageable than it would be otherwise.

Refs #12600

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13076** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13076" target="_blank">Graphite</a> 
    * **PR #13077** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13077" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/13077?utm_source=stack-comment).

**Comment by abhinav** [comment]
image

File: pkg/backend/filestate/backend.go [hunk 1]
```diff
@@ -402,27 +491,34 @@ func (b *localBackend) Upgrade(ctx context.Context) error {
...
+
+// upgradeStack upgrades a single stack to use the provided projectReferenceStore.
+func (b *localBackend) upgradeStack(
+	ctx context.Context,
+	newStore *projectReferenceStore,
+	project tokens.Name,
+	old *localBackendReference,
+) error {
+	contract.Requiref(old.project == "", "old.project", "must be empty")
 	if project == "" {
```
Comment (Frassle) [inline pkg/backend/filestate/backend.go:522]: Should we just make this a contract.Requires now? Make it invalid to call upgradeStack with project=""

**Review by Frassle — APPROVED** [review]

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
image

image


Will retry

**Comment by abhinav** [comment]
bors retry

**Pull request merged**